### PR TITLE
Add generic plugin test runner script

### PR DESCRIPTION
**Pull request by ehelms**
Plugins like Katello and foreman_puppet have their own wrapper scripts or Rakefile customizations to make it easier to run individual test files without manually switching to the Foreman directory and setting up cross-project load paths. This script generalizes that pattern for all plugins.

Supports running from the Foreman checkout, a plugin checkout, or anywhere on PATH with automatic Foreman/plugin discovery. Single-file runs use direct ruby invocation for faster feedback; full-suite runs delegate to the plugin's existing rake tasks.

I am submitting this as a replacement for [ktest](https://github.com/theforeman/puppet-katello_devel/blob/master/templates/ktest.sh.erb) to be available for any plugin, and support different paths of usage. And instead of this living inside our development deployment tooling, to live as a script in the main repository and thus allow it's usage from more places.

In the use case of `foremanctl`, the development deployer would copy the script to the users local bin directory to make it broadly available within the environment.

### CODE REVIEW

**Comment by adamruzicka** [comment]
Every now and then, I need to run an ad-hoc plugin rake task (usually `foreman_theme_satellite:validate_docs`), could this be extended to be able to run arbitrary rake tasks or would that be out of scope?

**Comment by ehelms** [comment]
> Every now and then, I need to run an ad-hoc plugin rake task (usually `foreman_theme_satellite:validate_docs`), could this be extended to be able to run arbitrary rake tasks or would that be out of scope?

I think that's fair. Do you think we can count on the pattern always including `:` to differentiate rake tasks from test files? Or do you think it will need an option like `--rake` ?

**Comment by adamruzicka** [comment]
> Or do you think it will need an option like --rake ?

I'd prefer this to also cover cases like string extraction where the plugin-name-prefix is not there (it is `plugin:gettext[$plugin]`)

File: script/plugin-test [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  return 1
+}
+
+detect_plugin_name() {
+  local dir="$1"
+  local gemspec
+  gemspec=$(compgen -G "$dir"/*.gemspec | head -1) || true
+  if [[ -n "$gemspec" ]]; then
+    basename "$gemspec" .gemspec
+  fi
```
Comment (adamruzicka) [inline script/plugin-test]: Here we'll run into foreman-tasks being silly with spaces and underscores. The gemspec is `foreman-tasks.gemspec`, but the rake task prefix is `foreman_tasks`.

File: script/plugin-test [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  OPTIONS       Extra arguments passed through to the test runner
+                (e.g. --name test_create).
+
+Examples:
+  plugin-test katello                                     # full suite via rake
+  plugin-test katello test/models/repo_test.rb            # single file
+  plugin-test test/models/repo_test.rb                    # from plugin checkout
+  plugin-test ../katello/test/models/repo_test.rb         # direct path from foreman
+  plugin-test test/models/repo_test.rb --name test_create
+  plugin-test foreman_theme_satellite:validate_docs       # run a rake task
```
Comment (adamruzicka) [inline script/plugin-test]: Could we `s/plugin-test/\$0` everywhere so the script can be freely renamed? Alternatively, use a more descriptive name like `foreman-plugin-test`?

File: script/plugin-test [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+EOF
+  exit 1
+}
+
+find_foreman() {
+  local dir="$1"
+  for candidate in "$dir" "$dir/foreman" "$dir/../foreman" "$dir/../../foreman"; do
+    if [[ -f "$candidate/config/application.rb" ]]; then
+      (cd "$candidate" && pwd)
```
Comment (adamruzicka) [inline script/plugin-test]: _Personal preference, feel free to ignore_
```suggestion
      readlink -f "$candidate"
```

File: script/plugin-test [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+      return 0
+    fi
+  done
+  return 1
+}
+
+detect_plugin_name() {
+  local dir="$1"
+  local gemspec
+  gemspec=$(compgen -G "$dir"/*.gemspec | head -1) || true
```
Comment (adamruzicka) [inline script/plugin-test]: _Personal preference, feel free to ignore_
I had to look up what compgen does, wouldn't this work as well?

```suggestion
  gemspec=$(ls -1 "$dir"/*.gemspec | head -1) || true
```

File: script/plugin-test [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+  local dir="$1"
+  local depth=0
+  while [[ "$dir" != "/" ]] && [[ $depth -lt 10 ]]; do
+    if compgen -G "$dir"/*.gemspec &>/dev/null; then
+      echo "$dir"
+      return 0
+    fi
+    dir=$(dirname "$dir")
+    depth=$((depth + 1))
+  done
```
Comment (adamruzicka) [inline script/plugin-test]:
```suggestion
  git -C "$dir" rev-parse --show-toplevel 2>/dev/null
```

File: script/foreman-plugin-test [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+  RAKE_PATH=$(bundle info rake --path) || {
+    echo "Error: Could not find rake in the bundle." >&2
+    exit 1
+  }
+
+  exec bundle exec ruby \
+    -I"$INCLUDE_PATHS" \
+    -I"${RAKE_PATH}/lib" \
+    "${TEST_FILES[@]}" \
+    "${OTHER_OPTS[@]}"
```
Comment (adamruzicka) [inline script/foreman-plugin-test]: I can't say I'm exactly happy about inventing yet another way of running the tests by modifying the include path. Considering we're limited to a single plugin, wouldn't this sort of work

```
exec bundle exec rake test:"$PLUGIN_NAME" "${OTHER_OPTS[@]}" TEST="$TEST_FILES[@]"
```

I'm a bit rusty with bashisms, so this might need some tweaking, but the point should be clear.

File: script/plugin-test (hunk 2 above)
Comment (ehelms) [reply script/plugin-test]: `foreman-plugin-test` sounds fine. Be great if we had some prefix that wasn't "foreman" since a hundred things share that bash-completion and name :grin:

File: script/foreman-plugin-test (hunk 6 above)
Comment (ehelms) [reply script/foreman-plugin-test]: I don't follow what the issue is, or what you are requesting a change for. Is it the use of `ruby` vs `rake` ?

Comment (adamruzicka) [reply script/foreman-plugin-test]: Test rake tasks can declare additional paths to be put on the load path. By invoking plugin tests without rake, we might be missing those additional paths. It is an edge case, true, but still.

Comment (ehelms) [reply script/foreman-plugin-test]: Those can fall back to the supply a rake task method. The original `ktest` tried to solve two problems:

1. make it easy to run Katello test's from it's home directory
2. speed up execution

The use of `rake` loads more of the environment than is needed. One option I have not explored is to try to have a special Rakefile for testing that's as fast as running ruby directly. Similar to how [Rakefile.dist](https://github.com/theforeman/foreman/blob/develop/Rakefile.dist) is much quicker for the tasks it handles because it avoids loading all of Rails.

Comment (adamruzicka) [reply script/foreman-plugin-test]: > Those can fall back to the supply a rake task method.

Fair

> One option I have not explored is to try to have a special Rakefile for testing

That's tempting

Comment (ehelms) [reply script/foreman-plugin-test]: The syntax (if it works) would be:

```
rake -f Rakefile.test test:single TEST=path/to/test.rb PLUGIN_PATH=/path/to/plugin
```

vs. this script:

```
#   1. From the Foreman checkout:  plugin-test katello test/models/foo_test.rb
#   2. From a plugin checkout:     plugin-test test/models/foo_test.rb
#   3. From anywhere (on PATH):    plugin-test katello test/models/foo_test.rb
#   4. With direct paths:          plugin-test ../katello/test/models/foo_test.rb
```

Comment (adamruzicka) [reply script/foreman-plugin-test]: When you put it like that, the script feels more ergonomic

**Comment by ehelms** [comment]
@adamruzicka updated from your feedback

File: script/foreman-plugin-test [hunk 7]
```diff
@@ -0,0 +1,198 @@
+#!/bin/bash
```
Comment (adamruzicka) [inline script/foreman-plugin-test]:
```suggestion
#!/usr/bin/env bash
```

Sadly, osx has ancient bash (3.2.57). Could we do this so that it can pick up bash from PATH, rather then explicitly choosing the system one?

It manifests itself like this
```
# foreman-plugin-test test/unit/smart_proxy_reference_test.rb
/Users/aruzicka/.local/bin/foreman-plugin-test: line 190: OTHER_OPTS[@]: unbound variable
```

**Review by adamruzicka — COMMENTED** [review]
One last comment, otherwise lgtm

**Review by adamruzicka — DISMISSED** [review]
This looks good, please squash the commits

**Comment by ehelms** [comment]
Green again -- @jeremylenz you might want to look at this and tell me what you think

File: script/foreman-plugin-test [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+                ":" is treated as a rake task (e.g. katello:reindex).
+  OPTIONS       Extra arguments passed through to the test runner
+                (e.g. --name test_create).
+
+Examples:
+  $cmd katello                                     # full suite via rake
+  $cmd katello test/models/repo_test.rb            # single file
+  $cmd test/models/repo_test.rb                    # from plugin checkout
+  $cmd ../katello/test/models/repo_test.rb         # direct path from foreman
+  $cmd test/models/repo_test.rb --name test_create
```
Comment (jeremylenz) [inline script/foreman-plugin-test]: When `--name` matches nothing, Minitest prints `0 tests, 0 assertions, 0 failures, 0 errors, 0 skips` and exits 0 — so foreman-plugin-test reports success even though nothing
  ran. May be worth fixing in the script by detecting the "0 tests" case and returning a non-zero exit / warning.

File: script/foreman-plugin-test [hunk 9]
```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env bash
+#
+# Run tests or rake tasks for a Foreman plugin.
+#
+# Works from four contexts:
+#   1. From the Foreman checkout:  foreman-plugin-test katello test/models/foo_test.rb
+#   2. From a plugin checkout:     foreman-plugin-test test/models/foo_test.rb
+#   3. From anywhere (on PATH):    foreman-plugin-test katello test/models/foo_test.rb
+#   4. With direct paths:          foreman-plugin-test ../katello/test/models/foo_test.rb
```
Comment (jeremylenz) [inline script/foreman-plugin-test]: Of course I tested with non-plugin tests too (in foreman directory) and it seems to work fine. With that in mind, maybe `ftest` (going off of ktest) ? or just `foreman-test` or `run-foreman-test`

File: script/foreman-plugin-test (hunk 8 above)
Comment (jeremylenz) [inline script/foreman-plugin-test]: `--name` only matches literal strings, so doesn't work with tests where the test is like
```rb
test 'child class should inherit loaders' do
```
because minitest generates a method name like
```rb
test_0001_child class should inherit loaders
```

If you enclose the string in a regex like
```bash
script/foreman-plugin-test test/unit/foreman/renderer/scope/base_test.rb \
--name "/child class should inherit loaders/"
```
it works. Not sure what the correct place is, but it'd be helpful to document that.

**Review by jeremylenz — COMMENTED** [review]
Thanks! Looks like an awesome replacement for ktest.

**Comment by jeremylenz** [comment]
This will be awesome for LLMs because there will be **one** way to run Ruby tests across the entire ecosystem. Is there a companion foremanctl PR for adding the script to bin?

**Review by ianballou — COMMENTED** [review]
I haven't tested it, but from a ktest functionality standpoint I'm happy with this PR. The biggest different is the sequential run of tests, which we can revisit in case speed is a concern.

**Comment by ehelms** [comment]
@jeremylenz Updated based on your feedback

**Review by jeremylenz — COMMENTED** [review]
It doesn't seem to be working right for me.

1. I seem to get all the output in one shot (after a long wait), rather than being able to follow along
2. I don't see the error you added:

```
$ script/foreman-test test/unit/foreman/renderer/scope/macros/host_template_test.rb --name "foo"
WARNING:  SET CONSTRAINTS can only be used in transaction blocks
Emptying /home/vagrant/foreman/jenkins/reports/unit
-- execute("SET CONSTRAINTS ALL DEFERRED;")
   -> 0.0011s

# Running tests with run options --name foo --seed 50466:

Writing XML reports to /home/vagrant/foreman/jenkins/reports/unit


Finished tests in 0.004709s, 0.0000 tests/s, 0.0000 assertions/s.


0 tests, 0 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by ehelms** [comment]
> * I seem to get all the output in one shot (after a long wait), rather than being able to follow along

That is due to having implemented ". May be worth fixing in the script by detecting the "0 tests" case and returning a non-zero exit / warning." But I think that's not a great path and am switching it back to streaming output.

**Review by jeremylenz — APPROVED** [review]
LGTM 👍 streams test output again.
